### PR TITLE
raftstore: return error when read index is dropped due to invaild leader

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

###  What have you changed?

When the leader id is `INVAILD_ID` in a replica, the read index request will be dropped by the raft group, see https://github.com/pingcap/raft-rs/blob/0632bcfc1ce4cd60812959758b7c8645f8d42458/src/raft.rs#L1968-L1979.
This PR adds a condition for returning an error directly in that situation, to fix endless waiting for the response of read index.

###  What is the type of the changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No.

###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -1884,6 +1884,9 @@ impl Peer {
         if self.is_merging() {
             return Err(box_err!("can not read index due to merge"));
         }
+        if !self.is_leader() && self.leader_id() == INVALID_ID {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: How did you find this bug? Better to cover it in schrodinger tests.

Comment (zhangjinpeng1987) [reply src/raftstore/store/peer.rs]: if the peer is not leader you can return error directly, why it must be equal to 0?

Comment (5kbpers) [reply src/raftstore/store/peer.rs]: > if the peer is not leader you can return error directly, why it must be equal to 0?

A replica will set `leader_id` to `INVALID_ID` when election timeout (which shows no leader).

File: tests/integrations/storage/test_raftkv.rs [hunk 2]
```diff
@@ -185,6 +188,66 @@ fn test_read_on_replica() {
...
+    request.mut_header().set_peer(p1.clone());
+    let (cb, rx) = make_cb(&request);
+    cluster
+        .sim
+        .rl()
+        .async_command_on_node(1, request.clone(), cb)
+        .unwrap();
+
+    let resp = rx.recv_timeout(time::Duration::from_millis(500)).unwrap();
+    assert!(resp.get_header().has_error());
```
Comment (overvenus) [inline tests/integrations/storage/test_raftkv.rs]: Please check whether the error message contains `can not read index due to no leader`.

File: src/raftstore/store/peer.rs (hunk 1 above)
Comment (overvenus) [reply src/raftstore/store/peer.rs]: Could you add a reference of how raft behaves on readindex if there is no leader?

**Comment by BusyJay** [comment]
Better add more check to

https://github.com/tikv/tikv/blob/80e35266be1bf8855da207876bbd9b3e669b0c1d/src/raftstore/store/peer.rs#L1942-L1958

instead.

**Comment by nolouch** [comment]
PTAL @BusyJay

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1941,12 +1958,29 @@ impl Peer {
             }
         }
 
+        // When a replica cannot detect any leader, `MsgReadIndex` will be dropped, which would
+        // cause a long time waiting for a read response. Then we should return an error directly
+        // in this situation.
+        if !self.is_leader() && self.leader_id() == INVALID_ID {
+            cmd_resp::bind_error(
```
Comment (Connor1996) [inline src/raftstore/store/peer.rs]: maybe update related metrics?

Comment (5kbpers) [reply src/raftstore/store/peer.rs]: > maybe update related metrics?

Should it be a type of `unsafe_read_index`?

Comment (Connor1996) [reply src/raftstore/store/peer.rs]: not for sure, add a new one? @overvenus What's your opinion?

Comment (overvenus) [reply src/raftstore/store/peer.rs]: What about `no_leader` or `follower_read_no_leader`? You can add it to `RaftInvalidProposeMetrics`.

Comment (5kbpers) [reply src/raftstore/store/peer.rs]: OK, I will add a `read_index_no_leader` there.

**Pull request merged**